### PR TITLE
Omit timeAdded from taint when empty

### PR DESCRIPTION
**Pull request by liggitt**
Fixes omitempty portion of https://github.com/kubernetes/kubernetes/issues/42394

### CODE REVIEW

**Comment by aveshagarwal** [comment]
@kubernetes/sig-scheduling-pr-reviews

**Comment by liggitt** [comment]
cc @smarterclayton @davidopp to verify we can do this in 1.7 without backcompat issues

**Comment by bsalamat** [comment]
Looks good. Thanks, @liggitt.

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller_test.go]: Why not add this into `metav1`, named `NowP() *Time`? So other developer can use it.

Comment (liggitt) [inline pkg/controller/daemon/daemoncontroller_test.go]: I'd rather not increase the exported API just yet

Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller_test.go]: Just grep the code, there're some similar logic to get "now" pointer. I'll create a PR to clear them up by "NowP()" :).

Comment (liggitt) [inline pkg/controller/daemon/daemoncontroller_test.go]: I think it dilutes the metav1 package... there's no `time.NowP()` function in the stdlib, for example

Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller_test.go]: I'm ok with that :)
Two lines to get "now" pointer seems boring :) ; just code style, no block comments to this PR :).

**Review by k82cn — APPROVED** [review]

**Comment by liggitt** [comment]
@k8s-bot non-cri e2e test this
@k8s-bot verify test this

**Comment by k82cn** [comment]
@liggitt , ping for rebase :).

**Comment by liggitt** [comment]
rebased, retagging

**Comment by k82cn** [comment]
I think we need to also update `pkg/controller/node/controller_utils.go` :).
 (https://github.com/kubernetes/kubernetes/pull/46671/commits/ba83a01c2d00d4de4da54fe59e95148010e41c27#diff-459b9981dc13bbbb313e652180644c3cL293 )

**Comment by k82cn** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by k82cn** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by k82cn** [comment]
@liggitt , would you help to rebase that?

**Comment by liggitt** [comment]
rebased/regenerated, PTAL

**Comment by liggitt** [comment]
fixed unit test

**Comment by liggitt** [comment]
flake on https://github.com/kubernetes/kubernetes/issues/49320
/retest

**Comment by k82cn** [comment]
need rebase :).

**Comment by liggitt** [comment]
@kubernetes/sig-scheduling-api-reviews @kubernetes/api-reviewers can I get approval on this before spending time keeping it rebased?

**Comment by k82cn** [comment]
@liggitt , this was approved, can you help to rebase ? so we can merge it :).

**Comment by liggitt** [comment]
rebased, will leave up to @kubernetes/sig-scheduling-api-reviews whether this is important for 1.8

File: staging/src/k8s.io/api/core/v1/types.go
Comment (gmarek) [inline staging/src/k8s.io/api/core/v1/types.go]: Isn't it a breaking API change in beta type?

Comment (liggitt) [reply staging/src/k8s.io/api/core/v1/types.go]: wire serialization remains compatible.

the old version encoded zero values to:
*  json: `timeAdded: null`
*  proto: 0x22 tag with zero length value

the new version omits zero values from the wire format

the new version successfully decodes zero values from the old version

the old version treats missing values as the zero value when decoding

**Comment by bsalamat** [comment]
As long as we are backward compatible, I am fine with the change.

/lgtm

**Comment by k82cn** [comment]
I'm fine to merge this in 1.9 :).

**Pull request merged**
